File: ios/Remotly/Views/Attachments.swift
// Photos attached to the composer: taken with the camera or picked from the library, downscaled on the
// phone, uploaded to the bridge at once, and appended to the next message as file paths the program in
// the pane reads (protocol §2 "Uploads"). Works the same for Claude Code, Codex and pi.
import FlowKit
import ImageIO
import PhotosUI
import SwiftUI
import UIKit

@MainActor
@Observable
final class Attachment: Identifiable {
    enum State: Equatable {
        case uploading
        case uploaded(path: String)
        case failed(String)
    }

    let id = UUID()
    let thumbnail: UIImage
    let jpeg: Data
    var state: State = .uploading

    init(thumbnail: UIImage, jpeg: Data) {
        self.thumbnail = thumbnail
        self.jpeg = jpeg
    }

    /// The path on the host once the bridge has stored the photo.
    var path: String? {
        if case .uploaded(let path) = state { return path }
        return nil
    }
}

// @MainActor: builds `Attachment`s (main-actor objects) and is only ever called from the composer.
@MainActor
enum ImagePrep {
    /// Long-edge cap: about what the models see anyway, and a few hundred KB instead of several MB.
    static let maxEdge: CGFloat = 1568
    static let quality: CGFloat = 0.85
    static let thumbnailEdge: CGFloat = 112

    /// Photo-library data (HEIC, JPEG, PNG …) → JPEG at most `maxEdge` on the long side, orientation baked in, metadata dropped.
    static func attachment(from data: Data) -> Attachment? {
        let sourceOptions = [kCGImageSourceShouldCache: false] as CFDictionary
        guard let source = CGImageSourceCreateWithData(data as CFData, sourceOptions) else { return nil }
        let options = [
            kCGImageSourceCreateThumbnailFromImageAlways: true,
            kCGImageSourceCreateThumbnailWithTransform: true,
            kCGImageSourceShouldCacheImmediately: true,
            kCGImageSourceThumbnailMaxPixelSize: maxEdge,
        ] as CFDictionary
        guard let cgImage = CGImageSourceCreateThumbnailAtIndex(source, 0, options) else { return nil }
        return attachment(from: UIImage(cgImage: cgImage))
    }

    /// A camera capture (already oriented by UIKit when drawn) → the same JPEG.
    static func attachment(from image: UIImage) -> Attachment? {
        let pixelWidth = image.size.width * image.scale
        let pixelHeight = image.size.height * image.scale
        guard pixelWidth > 0, pixelHeight > 0 else { return nil }
        let factor = min(1, maxEdge / max(pixelWidth, pixelHeight))
        let target = CGSize(width: floor(pixelWidth * factor), height: floor(pixelHeight * factor))
        guard let jpeg = render(image, into: target).jpegData(compressionQuality: quality) else { return nil }
        return Attachment(thumbnail: square(image, edge: thumbnailEdge), jpeg: jpeg)
    }

    /// Redraw at `size` in pixels on white (JPEG has no alpha), which also bakes the orientation in.
    private static func render(_ image: UIImage, into size: CGSize) -> UIImage {
        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        format.opaque = true
        return UIGraphicsImageRenderer(size: size, format: format).image { ctx in
            UIColor.white.setFill()
            ctx.fill(CGRect(origin: .zero, size: size))
            image.draw(in: CGRect(origin: .zero, size: size))
        }
    }

    /// Centre-cropped square for the chip.
    private static func square(_ image: UIImage, edge: CGFloat) -> UIImage {
        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        format.opaque = true
        let side = min(image.size.width, image.size.height)
        let scale = edge / max(side, 1)
        let drawSize = CGSize(width: image.size.width * scale, height: image.size.height * scale)
        let origin = CGPoint(x: (edge - drawSize.width) / 2, y: (edge - drawSize.height) / 2)
        return UIGraphicsImageRenderer(size: CGSize(width: edge, height: edge), format: format).image { _ in
            image.draw(in: CGRect(origin: origin, size: drawSize))
        }
    }
}

/// The system camera; `onImage(nil)` on cancel.
struct CameraPicker: UIViewControllerRepresentable {
    let onImage: (UIImage?) -> Void

    func makeUIViewController(context: Context) -> UIImagePickerController {
        let picker = UIImagePickerController()
        picker.sourceType = .camera
        picker.cameraCaptureMode = .photo
        picker.delegate = context.coordinator
        return picker
    }

    func updateUIViewController(_ uiViewController: UIImagePickerController, context: Context) {}

    func makeCoordinator() -> Coordinator { Coordinator(onImage: onImage) }

    @MainActor
    final class Coordinator: NSObject, @preconcurrency UIImagePickerControllerDelegate, UINavigationControllerDelegate {
        let onImage: (UIImage?) -> Void

        init(onImage: @escaping (UIImage?) -> Void) {
            self.onImage = onImage
        }

        func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey: Any]) {
            onImage(info[.originalImage] as? UIImage)
        }

        func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
            onImage(nil)
        }
    }
}

/// Thumbnails above the composer: a spinner while uploading, a red retry face after a failure, ✕ to drop.
struct AttachmentChips: View {
    let attachments: [Attachment]
    let onRemove: (Attachment) -> Void
    let onRetry: (Attachment) -> Void

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 10) {
                ForEach(attachments) { attachment in
                    ZStack(alignment: .topTrailing) {
                        Image(uiImage: attachment.thumbnail)
                            .resizable()
                            .scaledToFill()
                            .frame(width: 56, height: 56)
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                            .overlay { overlay(for: attachment) }
                            .onTapGesture { if case .failed = attachment.state { onRetry(attachment) } }
                            .accessibilityLabel(label(for: attachment))
                        Button { onRemove(attachment) } label: {
                            Image(systemName: "xmark.circle.fill")
                                .font(.body)
                                .foregroundStyle(.white, .black.opacity(0.65))
                        }
                        .offset(x: 7, y: -7)
                        .accessibilityLabel("Remove photo")
                    }
                    .padding(.top, 7)
                    .padding(.trailing, 7)
                }
            }
            .padding(.horizontal, 10)
        }
    }

    @ViewBuilder
    private func overlay(for attachment: Attachment) -> some View {
        switch attachment.state {
        case .uploading:
            ZStack {
                Color.black.opacity(0.35)
                ProgressView().tint(.white)
            }
            .clipShape(RoundedRectangle(cornerRadius: 8))
        case .failed:
            ZStack {
                Color.red.opacity(0.45)
                Image(systemName: "arrow.clockwise").foregroundStyle(.white)
            }
            .clipShape(RoundedRectangle(cornerRadius: 8))
        case .uploaded:
            EmptyView()
        }
    }

    private func label(for attachment: Attachment) -> String {
        switch attachment.state {
        case .uploading: return "Photo, uploading"
        case .uploaded: return "Photo, ready to send"
        case .failed(let message): return "Photo, upload failed: \(message). Tap to retry"
        }
    }
}

File: android/app/src/main/kotlin/com/inferenceaftermath/remotly/ui/Attachments.kt
// Photos attached to the composer: taken with the camera or picked from the photo picker, downscaled on the
// phone, uploaded to the bridge at once, and appended to the next message as file paths the program in the
// pane reads (protocol §2 "Uploads"). Works the same for Claude Code, Codex and pi.
package com.inferenceaftermath.remotly.ui

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageDecoder
import android.net.Uri
import android.os.Build
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.dp
import java.io.ByteArrayOutputStream
import java.io.File
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

/** One photo in the composer; [path] is set once the bridge has stored it. */
class Attachment(val thumbnail: Bitmap, val jpeg: ByteArray) {
    sealed interface State {
        data object Uploading : State
        data class Uploaded(val path: String) : State
        data class Failed(val message: String) : State
    }

    var state: State by mutableStateOf(State.Uploading)
    val path: String? get() = (state as? State.Uploaded)?.path
}

object ImagePrep {
    /** Long-edge cap: about what the models see anyway, and a few hundred KB instead of several MB. */
    const val MAX_EDGE = 1568
    const val QUALITY = 85
    const val THUMB_EDGE = 112

    /** Decode [uri] near MAX_EDGE (orientation applied on API 28+), re-encode as JPEG, drop metadata. */
    fun prepare(context: Context, uri: Uri): Attachment? = decode(context, uri)?.let { prepare(it) }

    fun prepare(bitmap: Bitmap): Attachment {
        val scaled = fit(bitmap, MAX_EDGE)
        val out = ByteArrayOutputStream()
        scaled.compress(Bitmap.CompressFormat.JPEG, QUALITY, out)
        return Attachment(square(scaled, THUMB_EDGE), out.toByteArray())
    }

    /** A private file for the camera app to write into (handed over through the FileProvider). */
    fun cameraFile(context: Context): File {
        val dir = File(context.cacheDir, "camera").apply { mkdirs() }
        return File(dir, "capture-${System.currentTimeMillis()}.jpg")
    }

    private fun decode(context: Context, uri: Uri): Bitmap? = try {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            ImageDecoder.decodeBitmap(ImageDecoder.createSource(context.contentResolver, uri)) { decoder, info, _ ->
                val longest = max(info.size.width, info.size.height)
                if (longest > MAX_EDGE) decoder.setTargetSampleSize(max(1, longest / MAX_EDGE))
                decoder.allocator = ImageDecoder.ALLOCATOR_SOFTWARE
                decoder.isMutableRequired = false
            }
        } else {
            val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
            context.contentResolver.openInputStream(uri)?.use { BitmapFactory.decodeStream(it, null, bounds) }
            val longest = max(bounds.outWidth, bounds.outHeight)
            val opts = BitmapFactory.Options().apply { inSampleSize = max(1, longest / MAX_EDGE) }
            context.contentResolver.openInputStream(uri)?.use { BitmapFactory.decodeStream(it, null, opts) }
        }
    } catch (e: Exception) {
        null
    }

    private fun fit(b: Bitmap, maxEdge: Int): Bitmap {
        val longest = max(b.width, b.height)
        if (longest <= maxEdge) return b
        val f = maxEdge.toFloat() / longest
        return Bitmap.createScaledBitmap(b, (b.width * f).roundToInt().coerceAtLeast(1), (b.height * f).roundToInt().coerceAtLeast(1), true)
    }

    /** Centre-cropped square for the chip. */
    private fun square(b: Bitmap, edge: Int): Bitmap {
        val side = min(b.width, b.height)
        val cropped = Bitmap.createBitmap(b, (b.width - side) / 2, (b.height - side) / 2, side, side)
        return Bitmap.createScaledBitmap(cropped, edge, edge, true)
    }
}

/** Thumbnails above the composer: a spinner while uploading, a red retry face after a failure, ✕ to drop. */
@Composable
fun AttachmentChips(attachments: List<Attachment>, onRemove: (Attachment) -> Unit, onRetry: (Attachment) -> Unit) {
    Row(
        Modifier.fillMaxWidth().horizontalScroll(rememberScrollState()).padding(horizontal = 10.dp, vertical = 2.dp),
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        for (a in attachments) {
            Box(Modifier.size(64.dp)) {
                Box(
                    Modifier.size(56.dp).align(Alignment.BottomStart).clip(RoundedCornerShape(8.dp))
                        .clickable(enabled = a.state is Attachment.State.Failed) { onRetry(a) },
                ) {
                    Image(a.thumbnail.asImageBitmap(), contentDescription = "Photo", contentScale = ContentScale.Crop, modifier = Modifier.fillMaxSize())
                    when (val s = a.state) {
                        Attachment.State.Uploading -> Box(Modifier.fillMaxSize().background(Color.Black.copy(alpha = 0.35f)), contentAlignment = Alignment.Center) {
                            CircularProgressIndicator(Modifier.size(22.dp), color = Color.White, strokeWidth = 2.dp)
                        }
                        is Attachment.State.Failed -> Box(Modifier.fillMaxSize().background(Color.Red.copy(alpha = 0.45f)), contentAlignment = Alignment.Center) {
                            Icon(Icons.Default.Refresh, contentDescription = "Upload failed: ${s.message}. Tap to retry", tint = Color.White)
                        }
                        is Attachment.State.Uploaded -> Unit
                    }
                }
                Icon(
                    Icons.Default.Close, contentDescription = "Remove photo", tint = Color.White,
                    modifier = Modifier.align(Alignment.TopEnd).size(20.dp).clip(RoundedCornerShape(10.dp))
                        .background(Color.Black.copy(alpha = 0.65f)).clickable { onRemove(a) }.padding(2.dp),
                )
            }
        }
    }
}
